8259842: Remove Result cache from StringCoding

cl4es opened the pull request:
The `StringCoding.resultCached` mechanism is used to remove the allocation of a `StringCoding.Result` object on potentially hot paths used in some `String` constructors. Using a `ThreadLocal` has overheads though, and the overhead got a bit worse after JDK-8258596 which addresses a leak by adding a `SoftReference`.

This patch refactors much of the decode logic back into `String` and gets rid of not only the `Result` cache, but the `Result` class itself along with the `StringDecoder` class and cache.

Microbenchmark results:
```
Baseline

Benchmark                                           (charsetName)  Mode  Cnt    Score    Error   Units
decodeCharset                                            US-ASCII  avgt   15  193.043 ±  8.207   ns/op
decodeCharset:·gc.alloc.rate.norm                        US-ASCII  avgt   15  112.009 ±  0.001    B/op
decodeCharset                                          ISO-8859-1  avgt   15  164.580 ±  6.514   ns/op
decodeCharset:·gc.alloc.rate.norm                      ISO-8859-1  avgt   15  112.009 ±  0.001    B/op
decodeCharset                                               UTF-8  avgt   15  328.370 ± 18.420   ns/op
decodeCharset:·gc.alloc.rate.norm                           UTF-8  avgt   15  224.019 ±  0.002    B/op
decodeCharset                                               MS932  avgt   15  328.870 ± 20.056   ns/op
decodeCharset:·gc.alloc.rate.norm                           MS932  avgt   15  232.020 ±  0.002    B/op
decodeCharset                                          ISO-8859-6  avgt   15  193.603 ± 12.305   ns/op
decodeCharset:·gc.alloc.rate.norm                      ISO-8859-6  avgt   15  112.010 ±  0.001    B/op
decodeCharsetName                                        US-ASCII  avgt   15  209.454 ±  9.040   ns/op
decodeCharsetName:·gc.alloc.rate.norm                    US-ASCII  avgt   15  112.009 ±  0.001    B/op
decodeCharsetName                                      ISO-8859-1  avgt   15  188.234 ±  7.533   ns/op
decodeCharsetName:·gc.alloc.rate.norm                  ISO-8859-1  avgt   15  112.009 ±  0.001    B/op
decodeCharsetName                                           UTF-8  avgt   15  399.463 ± 12.437   ns/op
decodeCharsetName:·gc.alloc.rate.norm                       UTF-8  avgt   15  224.019 ±  0.003    B/op
decodeCharsetName                                           MS932  avgt   15  358.839 ± 15.385   ns/op
decodeCharsetName:·gc.alloc.rate.norm                       MS932  avgt   15  208.017 ±  0.003    B/op
decodeCharsetName                                      ISO-8859-6  avgt   15  162.570 ±  7.090   ns/op
decodeCharsetName:·gc.alloc.rate.norm                  ISO-8859-6  avgt   15  112.009 ±  0.001    B/op
decodeDefault                                                 N/A  avgt   15  316.081 ± 11.101   ns/op
decodeDefault:·gc.alloc.rate.norm                             N/A  avgt   15  224.019 ±  0.002    B/op

Patched:
Benchmark                                           (charsetName)  Mode  Cnt    Score    Error   Units
decodeCharset                                            US-ASCII  avgt   15  159.153 ±  6.082   ns/op
decodeCharset:·gc.alloc.rate.norm                        US-ASCII  avgt   15  112.009 ±  0.001    B/op
decodeCharset                                          ISO-8859-1  avgt   15  134.433 ±  6.203   ns/op
decodeCharset:·gc.alloc.rate.norm                      ISO-8859-1  avgt   15  112.009 ±  0.001    B/op
decodeCharset                                               UTF-8  avgt   15  297.234 ± 21.654   ns/op
decodeCharset:·gc.alloc.rate.norm                           UTF-8  avgt   15  224.019 ±  0.002    B/op
decodeCharset                                               MS932  avgt   15  311.583 ± 16.445   ns/op
decodeCharset:·gc.alloc.rate.norm                           MS932  avgt   15  208.018 ±  0.002    B/op
decodeCharset                                          ISO-8859-6  avgt   15  156.216 ±  6.522   ns/op
decodeCharset:·gc.alloc.rate.norm                      ISO-8859-6  avgt   15  112.010 ±  0.001    B/op
decodeCharsetName                                        US-ASCII  avgt   15  180.752 ±  9.411   ns/op
decodeCharsetName:·gc.alloc.rate.norm                    US-ASCII  avgt   15  112.010 ±  0.001    B/op
decodeCharsetName                                      ISO-8859-1  avgt   15  156.170 ±  8.003   ns/op
decodeCharsetName:·gc.alloc.rate.norm                  ISO-8859-1  avgt   15  112.010 ±  0.001    B/op
decodeCharsetName                                           UTF-8  avgt   15  370.337 ± 22.199   ns/op
decodeCharsetName:·gc.alloc.rate.norm                       UTF-8  avgt   15  224.019 ±  0.001    B/op
decodeCharsetName                                           MS932  avgt   15  312.589 ± 15.067   ns/op
decodeCharsetName:·gc.alloc.rate.norm                       MS932  avgt   15  208.018 ±  0.002    B/op
decodeCharsetName                                      ISO-8859-6  avgt   15  173.205 ±  9.647   ns/op
decodeCharsetName:·gc.alloc.rate.norm                  ISO-8859-6  avgt   15  112.009 ±  0.001    B/op
decodeDefault                                                 N/A  avgt   15  303.459 ± 16.452   ns/op
decodeDefault:·gc.alloc.rate.norm                             N/A  avgt   15  224.019 ±  0.001    B/op
```

Most variants improve. There's a small added overhead in `String charsetName` variants for some charsets such as `ISO-8859-6` that benefited slightly from the `StringDecoder` cache due avoiding a lookup, but most variants are not helped by this cache and instead see a significant gain from skipping that step. `Charset` variants don't need a lookup and improve across the board.

Another drawback is that we need to cram more logic into `String` to bypass the `StringCoding.Result` indirection - but getting rid of two commonly used `ThreadLocal` caches and most cases getting a bit better raw throughput in the process I think more than enough makes up for that.

Testing: tier1-4

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8259842](https://bugs.openjdk.java.net/browse/JDK-8259842): Remove Result cache from StringCoding


### Reviewers
 * [Naoto Sato](https://openjdk.java.net/census#naoto) (@naotoj - **Reviewer**) ⚠️ Review applies to 2143cb3e59e1062ea8465720714dbd6cd17edafc
 * [Peter Levart](https://openjdk.java.net/census#plevart) (@plevart - **Reviewer**) ⚠️ Review applies to ba279a7e9ea840c3e9539bbfe3aaccd93f0e2748
 * [Roger Riggs](https://openjdk.java.net/census#rriggs) (@RogerRiggs - **Reviewer**) ⚠️ Review applies to 2143cb3e59e1062ea8465720714dbd6cd17edafc


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/2102/head:pull/2102`
`$ git checkout pull/2102`

RogerRiggs commented:
Interesting, I was/am in the middle of converting Result to be a Valhalla primitive class to reduce the memory pressure and had written some new jmh tests too.

cl4es commented:
> Interesting, I was/am in the middle of converting Result to be a Valhalla primitive class to reduce the memory pressure and had written some new jmh tests too.
> And to reduce the dependency on ThreadLocals.

Ok, I expect that would end up at similar performance while retaining the separation of concerns. But this way we're not dependent on valhalla to get rid of the TLs.

I'd be happy to add more JMH tests here. I expected this area to already have some, but it seems all the micros added during the work on compact strings work in JDK 9 are unaccounted for.

naotoj commented on src/java.base/share/classes/java/lang/String.java:
```diff
@@ -516,14 +533,490 @@ public String(byte bytes[], int offset, int length, String charsetName)
+            this.coder = "".coder;
+            return;
+        }
+        if (charset == UTF_8) {
```
The constructor is getting big. Might be better to keep the original private methods (decodeASCII/Latin1/UTF8) for readability.

cl4es replied:
Since we're calculating two final values, that split was what necessitated a `Result` object. Until valhalla I don't think there's a way to get rid of the performance cost here without putting the bulk of the logic into the constructor.

Refactoring out some of the logic to utility methods could be a performance neutral way to cut down the complexity, though. E.g.:
```
char c = (char)((b1 << 12) ^
                                                (b2 <<  6) ^
                                                (b3 ^
                                                 (((byte) 0xE0 << 12) ^
                                                  ((byte) 0x80 <<  6) ^
                                                  ((byte) 0x80 <<  0))));
                                if (Character.isSurrogate(c)) {
                                    putChar(dst, dp++, REPL);
                                } else {
                                    putChar(dst, dp++, c);
                                }
```
could be reasonably factored out and reduced to something like:
```
                                putChar(dst, dp++, StringCoding.decode3(b1, b2, b3));
```
I've refrained from refurbishing too much, though.

RogerRiggs replied:
I don't think you need to inline quite so much.  Once the determination has been made about whether the result is Latin1 or UTF16 it calls separate methods anyway.  For example, after calling hasNegatives() it is known what coding is needed.
Then call out to a method returns the byte array.

RogerRiggs replied:
(Also, every email is very long because it includes the performance data; next time please put the performance data in a separate comment, not the PR).

naotoj reviewed (commented):
Nit: copyright year in System.java is 2021.

plevart commented:
Hi Claes,
This is quite an undertaking in re-factoring!
I think that decoding logic that you produced can still be kept in StringCoding class though. The private constructor that you created for UTF-8 decoding is unnecessary. The logic can be kept in a static method and then the String instance constructed in it from the value and coder. The only public constructor that remains is then the one taking a Charset parameter. I took your code and moved it back to StringCoding while for the mentioned constructor I tried the following trick which could work (I haven't tested yet with JMH): if you pass an instance of newly constructed object down to a method that is inlined into the caller and that method does not pass that object to any other methods, JIT will eliminate the heap allocation, so I suspect that you can pass results from the called method back in that object and still avoid allocation...

https://github.com/plevart/jdk/commit/691600e3789a4c2b52fae921be87d0d7affa6f0f

WDYT?

cl4es commented:
> WDYT?

I get that the approach I took got a bit messy, but I've just spent some time cleaning it up. Please have a look at the latest, which outlines much of the logic and consolidates the replace/throw logic in the UTF8 decode paths. I've checked it does not regress on the micro, and I think the overall state of the code now isn't much messier than the original code.

plevart commented:
Some common logic is now extracted into methods. That's good. But much of the decoding logic still remains in String. I still think all of static methods can be moved to StringCoding directly as they are now while the private UTF-8 decoding constructor can be replaced with static method and moved to StringCoding. The only problem might be the public String constructor taking Charset parameter. But even here, passing a newly constructed mutable object to the method can be used to return multiple values from the method while relying on JIT to eliminate object allocation.
Do you think this code belongs more to String than to StringCoding?

cl4es commented:
> Do you think this code belongs more to String than to StringCoding?

I consider StringCoding an implementation detail of String, so I'm not sure there's (much) value in maintaining the separation of concern if it is cause for a performance loss. While encapsulating and separating concerns is a fine purpose I think the main purpose served by StringCoding is to resolve some bootstrap issues: String is one of the first classes to be loaded and eagerly pulling in dependencies like ThreadLocal and Charsets before bootstrapping leads to all manner of hard to decipher issues (yes - I've tried :-)).

plevart commented:
This looks much better now. Maybe just one small suggestion: `java.lang.StringCoding#lookupCharset` is used in two places and in both places the same exception handling/rethrowing logic is wrapped around the invocation. So you could move that logic into the lookupCharset method which would simplify call sites. You could even get rid of String.lookupCharset method that way.

plevart commented:
When you combine the logic of String.lookupCharset:

```
    private static Charset lookupCharset(String charsetName)
            throws UnsupportedEncodingException {
        Objects.requireNonNull(charsetName);
        try {
            Charset cs = StringCoding.lookupCharset(charsetName);
            if (cs == null) {
                throw new UnsupportedEncodingException(charsetName);
            }
            return cs;
        } catch (IllegalCharsetNameException ics) {
            throw new UnsupportedEncodingException(charsetName);
        }
    }
```

... and StringCoding.lookupCharset:

```
    static Charset lookupCharset(String csn) {
        if (Charset.isSupported(csn)) {
            try {
                return Charset.forName(csn);
            } catch (UnsupportedCharsetException x) {
                throw new Error(x);
            }
        }
        return null;
    }

```

...you get this:

```
    private static Charset lookupCharset(String charsetName)
            throws UnsupportedEncodingException {
        Objects.requireNonNull(charsetName);
        try {
            Charset cs;
            inner: {
                if (Charset.isSupported(charsetName)) {
                    try {
                        cs = Charset.forName(charsetName);
                        break inner;
                    } catch (UnsupportedCharsetException x) {
                        throw new Error(x);
                    }
                }
                cs = null;
            }
            if (cs == null) {
                throw new UnsupportedEncodingException(charsetName);
            }
            return cs;
        } catch (IllegalCharsetNameException ics) {
            throw new UnsupportedEncodingException(charsetName);
        }
    }

```

...and that can be simplified to this:

```
    static Charset lookupCharset(String csn) throws UnsupportedEncodingException {
        Objects.requireNonNull(csn);
        try {
            return Charset.forName(csn);
        } catch (UnsupportedCharsetException | IllegalCharsetNameException x) {
            throw new UnsupportedEncodingException(csn);
        }
    }
```

which has an additional benefit that it only performs one lookup (Charset.forName) instead of two (Charset.isSupported & Charset.forName)...

cl4es commented:
@plevart: I simplified the lookup logic based on your suggestion. Removed some unreachable paths in and simplified  `StringCoding.encode(String, byte, byte[])` as a result.

Simplifying to one lookup speeds up `decodeCharsetName` cases a notch:
```
before:
decodeCharsetName          UTF-8  avgt   15  370.337 ± 22.199  ns/op
after:
decodeCharsetName          UTF-8  avgt   15  335.971 ± 15.894  ns/op
```

plevart reviewed (approved):
This looks good.
Are you planning to do similar things for encoding too?

plevart commented:
I already approved the changes and they are OK. Maybe for a followup: just noticing after the fact that logic for `newStringUTF8NoRepl(....)` vs. `new String(...., StringCoding.UTF_8)` differ in handling unmappabale characters, which is by the spec, but also the constructor contains special handling of input that only contains non-negative bytes:

```
        if (charset == UTF_8) {
            if (COMPACT_STRINGS && !StringCoding.hasNegatives(bytes, offset, length)) {
                this.value = Arrays.copyOfRange(bytes, offset, offset + length);
                this.coder = LATIN1;
                return;
```

...while `newStringUTF8NoRepl(....)` does not contain this optimization. I guess ZipCoder could benefit from that optimization too since paths are mostly ASCII only. So WDYT of this additional simplification/consolidation of UTF-8 decoding:

https://github.com/plevart/jdk/commit/0b8b12c998e4ed451588442205ebe8f7423db7d8

cl4es commented:
> `newStringUTF8NoRepl(....)` does not contain this optimization.

Good catch: this optimization was in the original code for `newStringNoRepl` but not `newStringUTF8NoRepl`. I'll put it back to avoid a regression, but this time into `newStringUTF8NoRepl` so that both paths get the optimization.

plevart reviewed (approved):
This looks good, Claes.

RogerRiggs reviewed (commented):
The large number of package exposed methods in StringCoding is ugly and makes the code harder to maintain.
Can the code duplication of UTF8 in the String constructors be reduced?
It would be cleaner to move all of the remaining StringCoding methods into String and make them private again. Reading the code now requires quite a bit of cross referencing and the invariants are hard to verify.

RogerRiggs commented on src/java.base/share/classes/java/lang/StringCoding.java:
```diff
@@ -25,495 +25,21 @@
-    }
+    static final Charset ISO_8859_1 = sun.nio.cs.ISO_8859_1.INSTANCE;
+    static final Charset US_ASCII = sun.nio.cs.US_ASCII.INSTANCE;
+    static final Charset UTF_8 = sun.nio.cs.UTF_8.INSTANCE;
```
These could move to String also, if there is a benefit them being local.
Otherwise, the uses in String could refer directly to the INSTANCEs in the sun.nio.cs... classes.

RogerRiggs commented on src/java.base/share/classes/java/lang/StringCoding.java:
```diff
@@ -522,85 +48,18 @@ final String requestedCharsetName() {
      */
     private static native void err(String msg);
```
A separate cleanup could remove this unused method and the corresponding native code in StringCoding.c.

RogerRiggs commented on src/java.base/share/classes/java/lang/StringCoding.java:
```diff
@@ -522,85 +48,18 @@ final String requestedCharsetName() {
-    private static int implEncodeISOArray(byte[] sa, int sp,
+    public static int implEncodeISOArray(byte[] sa, int sp,
                                           byte[] da, int dp, int len) {
         int i = 0;
```
As a separate cleanup... 
If there isn't any value to these intrinsics being in StringCoder, they could also move to String
with the corresponding intrinsic references.

RogerRiggs reviewed (approved):
This looks very good. Thanks for the extra refactoring and consolidating of code.

cl4es commented:
Passed testing tiers 1-4

/integrate
